File: core/src/main/java/y2k/joyreactor/services/requests/TagsForUserRequest.java
package y2k.joyreactor.services.requests;

import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import rx.Observable;
import y2k.joyreactor.Image;
import y2k.joyreactor.Tag;
import y2k.joyreactor.common.ObservableUtils;
import y2k.joyreactor.http.HttpClient;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Created by y2k on 19/10/15.
 */
public class TagsForUserRequest {

    private TagImageRequest imageRequest = new TagImageRequest();
    private String username;

    public TagsForUserRequest(String username) {
        this.username = username;
    }

    public Observable<List<Tag>> request() {
        return ObservableUtils.create(() -> {
            Document document = HttpClient.getInstance().getDocument("http://joyreactor.cc/user/" + username);
            List<Tag> tags = new ArrayList<>();
            for (Element h : document.select(".sideheader")) {
                if ("Читает".equals(h.text())) {
                    for (Element a : h.parent().select("a"))
                        tags.add(createTag(a.text()));
                    break;
                }
            }
            Collections.sort(tags, (l, r) -> l.title.compareToIgnoreCase(r.title));
            return tags;
        });
    }

    private Tag createTag(String title) throws IOException {
        Tag tag = new Tag();
        tag.title = title;
        tag.image = new Image(imageRequest.request(title), 0, 0);
        return tag;
    }
}

File: core/src/main/kotlin/y2k/joyreactor/services/requests/TagsForUserRequest.kt
package y2k.joyreactor.services.requests

import y2k.joyreactor.common.async.CompletableFuture
import y2k.joyreactor.common.async.runAsync
import y2k.joyreactor.common.http.HttpClient
import y2k.joyreactor.common.platform.Platform
import y2k.joyreactor.model.Group
import y2k.joyreactor.model.Image
import java.util.*

/**
 * Created by y2k on 19/10/15.
 */
class TagsForUserRequest(
    private val httpClient: HttpClient,
    private val platform: Platform) {

    fun request(username: String): CompletableFuture<List<Group>> {
        val imageRequest = TagImageRequest(httpClient, platform)
        return runAsync {
            val document = httpClient.getDocument("http://joyreactor.cc/user/" + username)
            val tags = ArrayList<Group>()
            for (h in document.select(".sideheader")) {
                if ("Читает" == h.text()) {
                    for (a in h.parent().select("a")) {
                        val name = a.text()
                        tags.add(Group.makeTag(name, Image(imageRequest.request(name))))
                    }
                    break
                }
            }
            tags
        }
    }
}